knimin/lib/data_access.py

```diff
@@ -2759,6 +2767,155 @@ def delete_samples(self, ids):
+        """
+        sql = """SELECT EXISTS (SELECT 1 FROM pm.study_sample
+                                WHERE study_id = %s
+                                AND sample_id = %s)"""
+        return self._con.execute_fetchone(sql, [study_id, sample_id])[0]
+
+    def associate_samples_with_study(self, study_id, sample_ids):
+        """Creates associations between one or more samples and one study
+
+        Parameters
+        ----------
+        study_id : int
+            ID of the study
+        sample_ids : list of str
+            IDs of the samples
+
+        Raises
+        ------
+        ValueError
+            If the study ID does not exist
+            If the study and the sample are already associated
+        """
+        self._study_exists(study_id)
+        with TRN:
+            for sample_id in sample_ids:
+                if not self._sample_exists(sample_id):
+                    raise ValueError('Sample ID %s does not exist.'
+                                     % sample_id)
+                if self._study_sample_are_associated(study_id, sample_id):
+                    raise ValueError('Study ID %s and sample ID %s are already'
+                                     ' associated.' % (study_id, sample_id))
+                sql = """INSERT INTO pm.study_sample (study_id, sample_id)
+                         VALUES (%s, %s)"""
+                TRN.add(sql, [study_id, sample_id])
+
+    def dissociate_samples_from_study(self, study_id, sample_ids):
+        """Deletes the associations between one or more samples and one study
+
+        Parameters
+        ----------
+        study_id : int
+            ID of the study
+        sample_ids : list of str
+            IDs of the samples
+
+        Raises
+        ------
+        ValueError
+            If the study ID does not exist
+            If the study and the sample are not associated
+        """
+        self._study_exists(study_id)
+        with TRN:
+            for sample_id in sample_ids:
+                if not self._sample_exists(sample_id):
+                    raise ValueError('Sample ID %s does not exist.'
+                                     % sample_id)
+                if not self._study_sample_are_associated(study_id, sample_id):
+                    raise ValueError('Study ID %s and sample ID %s are not '
+                                     'associated.' % (study_id, sample_id))
+                sql = """DELETE FROM pm.study_sample
+                         WHERE study_id = %s AND sample_id = %s"""
+                TRN.add(sql, [study_id, sample_id])
+
+    def get_samples_associated_with_study(self, study_id):
+        """Retrieves IDs of all samples associated with a study (if any)
+
+        Parameters
+        ----------
+        study_id : int
+            ID of the study
+
+        Returns
+        ------
+        list of str
+            Sorted list of sample IDs associated with the study
+        """
+        self._study_exists(study_id)
```

Comment:
Hi @josenavas May I double-confirm with you, that do you want all SQL operations to be done using TRN, even if the operation does not write the database?